packages/components/CONTRIBUTING.md

```diff
@@ -4,12 +4,134 @@ Thank you for taking the time to contribute.
 
 The following is a set of guidelines for contributing to the `@wordpress/components` package to be considered in addition to the general ones described in our [Contributing Policy](/CONTRIBUTING.md).
 
-## Examples
+## Core principles
 
-Each component needs to include an example in its README.md file to demonstrate the usage of the component.
+Contributions to the `@wordpress/components` package should follow a set of core principles and technical requirements.
 
-These examples can be consumed automatically from other projects in order to visualize them in their documentation. To ensure these examples are extractable, compilable and renderable, they should be structured in the following way:
+This set of guidelines should apply especially to newly introduced components. It is, in fact, possible that some of the older component don't respect some of these guidelines for legacy/compat reasons.
 
--   It has to be included in a `jsx` code block.
--   It has to work out-of-the-box. No additional code should be needed to have working the example.
--   It has to define a React component called `My<ComponentName>` which renders the example (i.e.: `MyButton`). Examples for the Higher Order Components should define a `MyComponent<ComponentName>` component (i.e.: `MyComponentWithNotices`).
+### Compatibility
+
+The `@wordpress/components` package includes several components which are relied upon by many developers across different projects. It is, therefore, very important to avoid introducing breaking changes as much as possible.
+
+In these situations, an alternative approach is to "soft-deprecate" the given legacy API. This is achieved by removing traces of the API from the docs, while the code keeps supporing it in the background (with or without a warning)
+
+### Components structure
+
+The contents of the `@wordpress/components` package can be divided into three groups: UI Primitives, UI Blocks, and Utilities.
```

Comment:
The proposed structure isn't reflected in the code as of today (as noted below). It is confusing when you don't have a deep understanding of the subtle differences between UI primitives or UI blocks. At the end of the day, what is the most important is what gets exposed as public API. I don't think we ever want to expose utilities as they should rather be used to help with code reuse internally.

Reply:
I replied to a similar point made by @priethor in [this other comment](https://github.com/WordPress/gutenberg/pull/33960#discussion_r695093491). Let's move this conversation there